File: OtherLanguages/SnakeGame.go
// snake_game.go
// Author: Sanne Karibo
// Simple terminal Snake game in Go

package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"os"
	"os/exec"
	"runtime"
	"time"
)

const (
	width     = 64
	height    = 24
	maxLength = 100
)

type Point struct {
	x, y int
}

type Game struct {
	snake    []Point
	dir      int // 1=right,2=down,3=left,4=up
	food     Point
	score    int
	gameOver bool
}

func clearScreen() {
	switch runtime.GOOS {
	case "windows":
		cmd := exec.Command("cmd", "/c", "cls")
		cmd.Stdout = os.Stdout
		cmd.Run()
	default:
		fmt.Print("\033[2J\033[H")
	}
}

func (g *Game) init() {
	for i := 0; i < 5; i++ {
		g.snake = append(g.snake, Point{20 - i, 10})
	}
	g.dir = 1
	g.generateFood()
}

func (g *Game) generateFood() {
	rand.Seed(time.Now().UnixNano())
	for {
		x := rand.Intn(width)
		y := rand.Intn(height)
		if !g.contains(x, y) {
			g.food = Point{x, y}
			break
		}
	}
}

func (g *Game) contains(x, y int) bool {
	for _, p := range g.snake {
		if p.x == x && p.y == y {
			return true
		}
	}
	return false
}

func (g *Game) moveSnake() {
	head := g.snake[0]
	var newHead Point
	switch g.dir {
	case 1:
		newHead = Point{head.x + 1, head.y}
	case 2:
		newHead = Point{head.x, head.y + 1}
	case 3:
		newHead = Point{head.x - 1, head.y}
	case 4:
		newHead = Point{head.x, head.y - 1}
	}

	if newHead.x < 0 || newHead.x >= width || newHead.y < 0 || newHead.y >= height {
		g.gameOver = true
		return
	}
	if g.contains(newHead.x, newHead.y) {
		g.gameOver = true
		return
	}

	g.snake = append([]Point{newHead}, g.snake...)

	if newHead == g.food {
		g.score += 10
		g.generateFood()
		if len(g.snake) > maxLength {
			g.snake = g.snake[:maxLength]
		}
	} else {
		g.snake = g.snake[:len(g.snake)-1]
	}
}

func (g *Game) draw() {
	clearScreen()
	fmt.Printf("Score: %d\n", g.score)

	grid := make([][]rune, height)
	for i := range grid {
		grid[i] = make([]rune, width)
		for j := range grid[i] {
			grid[i][j] = ' '
		}
	}

	for i, p := range g.snake {
		if i == 0 {
			grid[p.y][p.x] = 'O'
		} else {
			grid[p.y][p.x] = 'o'
		}
	}
	grid[g.food.y][g.food.x] = 'X'

	for _, row := range grid {
		fmt.Println(string(row))
	}
}

func main() {
	game := Game{}
	game.init()

	reader := bufio.NewReader(os.Stdin)

	fmt.Println("Snake Game - Created by Sanne Karibo")
	fmt.Println("Use WASD keys + Enter to move. Press Enter to start...")
	reader.ReadString('\n')

	for !game.gameOver {
		game.draw()

		// Read user input for direction
		input, _ := reader.ReadString('\n')
		switch input {
		case "d\n", "D\n":
			if game.dir != 3 {
				game.dir = 1
			}
		case "s\n", "S\n":
			if game.dir != 4 {
				game.dir = 2
			}
		case "a\n", "A\n":
			if game.dir != 1 {
				game.dir = 3
			}
		case "w\n", "W\n":
			if game.dir != 2 {
				game.dir = 4
			}
		}

		game.moveSnake()
		time.Sleep(150 * time.Millisecond)
	}

	game.draw()
	fmt.Println("Game Over! Final Score:", game.score)
	fmt.Println("Press Enter to exit...")
	reader.ReadString('\n')
}

File: OtherLanguages/SnakeGame.exs
# snake_game.exs
# Author: Sanne Karibo
# Terminal snake game in Elixir

defmodule SnakeGame do
  @width 64
  @height 24
  @max_length 100

  defstruct snake: [], dir: :right, food: {0, 0}, score: 0, game_over: false

  def start do
    snake = Enum.map(0..4, fn i -> {20 - i, 10} end)
    food = generate_food(snake)
    game = %__MODULE__{snake: snake, food: food}
    IO.puts "Snake Game - Created by Sanne Karibo"
    IO.puts "Use WASD keys to move. Press Enter to start..."
    IO.gets("")

    loop(game)
  end

  defp loop(%{game_over: true, score: score}) do
    IO.puts "Game Over! Final Score: #{score}"
  end

  defp loop(game) do
    draw(game)
    input = IO.gets("") |> String.trim()
    game = update_direction(game, input)
    game = move_snake(game)
    :timer.sleep(150)
    loop(game)
  end

  defp generate_food(snake) do
    food = { :rand.uniform(@width) - 1, :rand.uniform(@height) - 1 }
    if food in snake do
      generate_food(snake)
    else
      food
    end
  end

  defp update_direction(game, input) do
    dir = game.dir
    cond do
      input == "d" and dir != :left -> %{game | dir: :right}
      input == "s" and dir != :up -> %{game | dir: :down}
      input == "a" and dir != :right -> %{game | dir: :left}
      input == "w" and dir != :down -> %{game | dir: :up}
      true -> game
    end
  end

  defp move_snake(%{game_over: true} = game), do: game

  defp move_snake(game) do
    [{hx, hy} | _] = game.snake
    new_head = case game.dir do
      :right -> {hx + 1, hy}
      :down -> {hx, hy + 1}
      :left -> {hx - 1, hy}
      :up -> {hx, hy - 1}
    end

    cond do
      out_of_bounds?(new_head) or new_head in game.snake ->
        %{game | game_over: true}

      new_head == game.food ->
        new_snake = [new_head | game.snake]
        new_food = generate_food(new_snake)
        %{game | snake: new_snake, food: new_food, score: game.score + 10}

      true ->
        new_snake = [new_head | Enum.slice(game.snake, 0..-2)]
        %{game | snake: new_snake}
    end
  end

  defp out_of_bounds?({x, y}) do
    x < 0 or x >= @width or y < 0 or y >= @height
  end

  defp draw(game) do
    IO.write("\e[H\e[2J") # clear screen
    IO.puts("Score: #{game.score}")
    for y <- 0..(@height - 1) do
      row =
        for x <- 0..(@width - 1) do
          cond do
            {x, y} == hd(game.snake) -> "O"
            {x, y} in tl(game.snake) -> "o"
            {x, y} == game.food -> "X"
            true -> " "
          end
        end
        |> Enum.join("")
      IO.puts(row)
    end
  end
end

SnakeGame.start()
